src/Dictionaries/HashedDictionary.h

```diff
@@ -203,6 +218,19 @@ class HashedDictionary final : public IDictionary
 
     void calculateBytesAllocated();
 
+    UInt64 getShard(UInt64 key) const
+    {
+        if constexpr (!sharded)
+            return 0;
+        return key % configuration.shards;
```

Comment:
When we take modulo of key without hashing it first, we can get issues with distribution.

Reply:
Yes, that was just to avoid one extra hash calculation, but OK, will add hash calculation here as well.

Reply:
Although fixing this, makes impossible to use `createBlockSelector`, and without this two changes it works 2.4x slower.